File: backend/services/import_/numbering.py
"""Derive positional clause numbers from tree position (F04, DD-02).

Clause numbers are a *projection* of structure, not stored source text: a node's
number is its decimal-outline path from the roots. Roots (parent_index None),
ordered by order_index, are "1", "2", …; their children "1.1", "1.2", …; a node
at depth N gets N dotted segments. Re-derived on every structural edit so a
promote/demote/move renumbers the subtree automatically.

This intentionally ignores the source's own letter/roman numbering (e.g. "(a)",
"(iv)") for v1 — decimal outline only. Letter/roman scheme preservation is a
later concern; positional decimals are unambiguous and sufficient for review.

Only `clause`-role nodes are numbered (DD-54): front-matter, signature_block,
appendix, and drafting_note are excluded from the clause tree and carry no
number. Non-clause siblings do not consume a position, so the operative tree
re-derives from the first real clause — fixing the spurious 1/2/3 the parser used
to stamp on the title page and recitals.

Block enumerated items (DD-98 / F03f) are also excluded: an `(a)`/`(A)`/`(i)`
item that sits as its own block under a clause keeps its literal marker as native
text and is addressed as "1.2.1(b)", never decimal-renumbered to "1.2.1.2". The
skip keys on the marker that is already in the node text (`is_block_enumerator`)
plus the import-set `enumerated` flag (which also catches a Word alpha/roman
auto-numbered item whose marker is generated, not in text). Like a non-clause
node, an enumerated item consumes no sibling position, so a real sub-clause
beside it still numbers correctly.
"""

from __future__ import annotations

import re
from collections import defaultdict

from backend.models.contract_tree import ParsedTree, TreeNode

# A leading parenthesised enumerator marker — "(a)", "(A)", "(iv)", "(II)". The
# marker may be glued to the text ("(A)The …" in real recitals), so no trailing
# space is required.
_ENUM_MARKER = re.compile(r"^\s*\(([A-Za-z]{1,4})\)")
_ROMAN_LABELS = frozenset(
    [
        "i",
        "ii",
        "iii",
        "iv",
        "v",
        "vi",
        "vii",
        "viii",
        "ix",
        "x",
        "xi",
        "xii",
        "xiii",
        "xiv",
        "xv",
        "xvi",
        "xvii",
        "xviii",
        "xix",
        "xx",
    ]
)


def is_block_enumerator(text: str) -> bool:
    """True when `text` opens with a block-enumerated-item marker: a single alpha
    letter or a roman numeral in parentheses. Curated to standard legal
    enumerators (single letter a–z, or roman i–xx) so a parenthetical word
    ("(the) …", "(or) …") is never mistaken for an enumerator (DD-98)."""
    m = _ENUM_MARKER.match(text)
    if m is None:
        return False
    label = m.group(1).lower()
    return (len(label) == 1 and label.isalpha()) or label in _ROMAN_LABELS


def derive_numbers(tree: ParsedTree) -> dict[int, str]:
    children: dict[int | None, list[TreeNode]] = defaultdict(list)
    for node in tree.nodes:
        children[node.parent_index].append(node)
    for siblings in children.values():
        siblings.sort(key=lambda n: n.order_index)

    numbers: dict[int, str] = {}

    def assign(parent: int | None, prefix: str) -> None:
        position = 0
        for node in children.get(parent, []):
            if node.role == "clause" and not (node.enumerated or is_block_enumerator(node.text)):
                position += 1
                number = f"{prefix}.{position}" if prefix else str(position)
                numbers[node.index] = number
                assign(node.index, number)
            else:
                # Non-clause or block-enumerated node: no number, no position
                # consumed. Recurse with the same prefix so any clause nested
                # beneath still numbers.
                assign(node.index, prefix)

    assign(None, "")
    return numbers


def _alpha(n: int) -> str:
    """1->a, 2->b, … 26->z, 27->aa (bijective base-26, lowercase)."""
    s = ""
    while n > 0:
        n, r = divmod(n - 1, 26)
        s = chr(97 + r) + s
    return s


def _roman(n: int) -> str:
    """Standard lowercase roman numeral for n>=1."""
    table = [
        (1000, "m"),
        (900, "cm"),
        (500, "d"),
        (400, "cd"),
        (100, "c"),
        (90, "xc"),
        (50, "l"),
        (40, "xl"),
        (10, "x"),
        (9, "ix"),
        (5, "v"),
        (4, "iv"),
        (1, "i"),
    ]
    out = ""
    for value, sym in table:
        while n >= value:
            out += sym
            n -= value
    return out


def format_enumerator(ordinal: int, enumerator_format: str | None) -> str:
    """The parenthesised marker glyph for the `ordinal`-th (1-based) item of a list
    in `enumerator_format` — "(a)" / "(A)" / "(i)" / "(I)". Empty for an unknown
    format. The glyph is derived from position, so it auto-renumbers (DD-99)."""
    if enumerator_format == "lowerLetter":
        return f"({_alpha(ordinal)})"
    if enumerator_format == "upperLetter":
        return f"({_alpha(ordinal).upper()})"
    if enumerator_format == "lowerRoman":
        return f"({_roman(ordinal)})"
    if enumerator_format == "upperRoman":
        return f"({_roman(ordinal).upper()})"
    if enumerator_format == "decimal":
        return f"({ordinal})"
    return ""


def derive_enumerators(tree: ParsedTree) -> dict[int, str]:
    """Marker glyph (e.g. "(a)") for every AUTO-NUMBERED enumerated node (those with
    an `enumerator_format`), derived from list position so it auto-renumbers on
    delete/insert (DD-99/F03f). The ordinal counts CONSECUTIVE enumerated siblings of
    the same format under one parent, restarting at 1 when the run starts, a
    non-enumerated sibling breaks the run, or the format changes — this is the
    restart-per-list-run behaviour real Word lists need (one source numId is reused
    across many separate lists). Literal-marker items (no `enumerator_format`, marker
    frozen in body) are intentionally excluded."""
    children: dict[int | None, list[TreeNode]] = defaultdict(list)
    for node in tree.nodes:
        children[node.parent_index].append(node)
    for siblings in children.values():
        siblings.sort(key=lambda n: n.order_index)

    markers: dict[int, str] = {}
    for siblings in children.values():
        count = 0
        prev_format: str | None = None
        for node in siblings:
            fmt = node.enumerator_format if node.enumerated else None
            if fmt is None:
                prev_format = None  # break the run
                continue
            count = count + 1 if fmt == prev_format else 1
            markers[node.index] = format_enumerator(count, fmt)
            prev_format = fmt
    return markers

File: frontend/app/lib/numbering.ts
// Re-derive positional outline numbers from a depth-ordered node list — mirrors
// the backend `derive_numbers` + `derive_enumerators` + `_number_for`. Used to
// renumber live when a node is promoted, demoted, or deleted so the number always
// follows the structure (DD-02), and so block enumerated items keep their derived
// "(a)"/"(1)" marker rather than being decimal-renumbered (F03f / DD-99).
//
// The numbering LEVEL of a decimal clause is its clause-ancestry depth (nearest
// preceding shallower numbered node = parent), computed by a stack walk, NOT the raw
// depth — so a depth gap (clause under a non-clause lead-in, or after a demote/move)
// does not emit a phantom `.0` level. An enumerated node consumes NO clause position:
// it is displayed as `<parent decimal> + (marker)` where the marker is derived from
// its ordinal among consecutive same-format enumerated siblings (restart per run),
// mirroring the backend exactly. A literal-marker enumerated item (no captured
// format — the "(a)" lives in its body text) is skipped entirely and carries no
// number, since its marker is already in the text.

export interface NumberingNode {
  depth: number;
  // True for a block enumerated item: a captured auto-numbered list format OR a
  // literal paren marker in the body (see isBlockEnumerator). Either way it is not
  // decimal-renumbered.
  enumerated: boolean;
  // Captured Word list format for an auto-numbered item; null for an ordinary clause
  // or a literal-marker item. Selects the marker glyph.
  enumeratorFormat: string | null;
}

const ROMAN_LABELS: ReadonlySet<string> = new Set([
  "i", "ii", "iii", "iv", "v", "vi", "vii", "viii", "ix", "x",
  "xi", "xii", "xiii", "xiv", "xv", "xvi", "xvii", "xviii", "xix", "xx",
]);

const ENUM_MARKER = /^\s*\(([A-Za-z]{1,4})\)/;

// Mirror of backend numbering.is_block_enumerator: text opens with a single alpha
// letter or a roman numeral in parens. Curated so a parenthetical word ("(or) …")
// is never mistaken for an enumerator (DD-98).
export function isBlockEnumerator(text: string): boolean {
  const m = ENUM_MARKER.exec(text);
  if (m === null) return false;
  const label = m[1].toLowerCase();
  return (label.length === 1 && /[a-z]/.test(label)) || ROMAN_LABELS.has(label);
}

// 1->a, 2->b, … 26->z, 27->aa (bijective base-26, lowercase). Mirrors backend _alpha.
function alpha(n: number): string {
  let s = "";
  let x = n;
  while (x > 0) {
    const r = (x - 1) % 26;
    x = Math.floor((x - 1) / 26);
    s = String.fromCharCode(97 + r) + s;
  }
  return s;
}

// Standard lowercase roman numeral for n>=1. Mirrors backend _roman.
function roman(n: number): string {
  const table: Array<[number, string]> = [
    [1000, "m"], [900, "cm"], [500, "d"], [400, "cd"], [100, "c"],
    [90, "xc"], [50, "l"], [40, "xl"], [10, "x"],
    [9, "ix"], [5, "v"], [4, "iv"], [1, "i"],
  ];
  let out = "";
  let x = n;
  for (const [value, sym] of table) {
    while (x >= value) {
      out += sym;
      x -= value;
    }
  }
  return out;
}

// The parenthesised marker glyph for the ordinal-th (1-based) item of a list in
// `fmt`. Mirrors backend format_enumerator (incl. the decimal branch, DD-99 amended).
export function formatEnumerator(ordinal: number, fmt: string | null): string {
  switch (fmt) {
    case "lowerLetter": return `(${alpha(ordinal)})`;
    case "upperLetter": return `(${alpha(ordinal).toUpperCase()})`;
    case "lowerRoman": return `(${roman(ordinal)})`;
    case "upperRoman": return `(${roman(ordinal).toUpperCase()})`;
    case "decimal": return `(${ordinal})`;
    default: return "";
  }
}

export function deriveNumbers(nodes: NumberingNode[]): string[] {
  const counters: number[] = [];
  const stack: Array<{ depth: number; number: string }> = []; // numbered-clause ancestors
  // Enumeration run state PER DEPTH (not one flat run): a list that resumes after a deeper
  // sub-list must continue, not restart. runFmt[d]/runCount[d] hold the open run at depth d;
  // returning to a shallower node drops the deeper entries, and a non-enumerated node at
  // depth d breaks the run there — mirroring backend derive_enumerators (group by sibling).
  // This makes (a)(A)(B)(b)(c) number correctly, and a re-indent (bringing a stray second
  // (a) up to its siblings' depth) re-derive it to (b).
  const runFmt: Array<string | null> = [];
  const runCount: number[] = [];

  return nodes.map((node) => {
    const d = node.depth;
    while (stack.length && stack[stack.length - 1].depth >= d) stack.pop();
    // Close runs deeper than this node.
    runFmt.length = d + 1;
    runCount.length = d + 1;

    if (node.enumerated) {
      if (node.enumeratorFormat === null) return ""; // literal marker — already in body
      const parentNumber = stack.length ? stack[stack.length - 1].number : "";
      if (runFmt[d] === node.enumeratorFormat) runCount[d] += 1;
      else {
        runCount[d] = 1;
        runFmt[d] = node.enumeratorFormat;
      }
      return `${parentNumber}${formatEnumerator(runCount[d], node.enumeratorFormat)}`;
    }

    // Ordinary clause: breaks any enumeration run at this depth.
    runFmt[d] = null;
    runCount[d] = 0;

    const level = stack.length; // numbering level = numbered-ancestry depth
    counters.length = level + 1; // drop any deeper counters
    for (let i = 0; i <= level; i++) if (counters[i] === undefined) counters[i] = 0;
    counters[level] += 1;
    const number = counters.slice(0, level + 1).join(".");
    stack.push({ depth: d, number });
    return number;
  });
}

// Re-derive each node's parent from the document-ordered (index, depth) sequence
// via a stack walk: a node's parent is the nearest preceding node at depth-1.
// This mirrors the backend tree_builder (parent = last_at_depth[depth-1]), so for
// untouched nodes it reproduces the original parent_index exactly, while honoring
// the operator's promote/demote (level ±) edits as real structural reparenting.
// Returns the parent node's `index` per row (null for a depth-0 root).
export function deriveParents(rows: { index: number; depth: number }[]): (number | null)[] {
  const stack: { index: number; depth: number }[] = [];
  return rows.map((r) => {
    while (stack.length && stack[stack.length - 1].depth >= r.depth) stack.pop();
    const parent = stack.length ? stack[stack.length - 1].index : null;
    stack.push(r);
    return parent;
  });
}
